### PR TITLE
API: fix str-accessor on CategoricalIndex

### LINKED ISSUE #23555: BUG: CategoricalIndex.str.extractall fails
Found a bug in a corner case while working on expanded test coverage in #23167:

```
>>> # the following three have the same result
>>> pd.Series(['a', 'b', 'aa'], dtype='category').str.extractall(r'(a)')
>>> pd.Series(['a', 'b', 'aa']).str.extractall(r'(a)')
>>> pd.Index(['a', 'b', 'aa']).str.extractall(r'(a)')
         0
  match
0 0      a
2 0      a
  1      a
>>> pd.Index(['a', 'b', 'aa'], dtype='category').str.extractall(r'(a)')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 2567, in extractall
    return str_extractall(self._orig, pat, flags=flags)
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 1012, in str_extractall
    is_mi = arr.index.nlevels > 1
AttributeError: 'CategoricalIndex' object has no attribute 'index'
```

So `Series`, categorical `Series` and `Index` work with `extractall`, but not `CategoricalIndex`.

### LINKED ISSUE #23556: BUG: many methods on CategoricalIndex.str are broken
This was also uncovered by #23167, but is a different error than #23555.

Basically, all methods calling `CategoricalIndex.str._wrap_result(result, use_codes=True)` will necessarily fail, e.g.:
```
>>> import pandas as pd
>>> pd.Index(['a', 'b', 'aa'], dtype='category').str.replace('a', 'c')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 2430, in replace
    return self._wrap_result(result)
  File "C:\ProgramData\Miniconda3\envs\pandas-dev\lib\site-packages\pandas\core\strings.py", line 1964, in _wrap_result
    result = take_1d(result, self._orig.cat.codes)
AttributeError: 'CategoricalIndex' object has no attribute 'cat'
```

This is because `self._orig` is the original `CategoricalIndex`, which does not have a `cat`-accessor.

### PR DESCRIPTION
**Pull request by h-vetinari**
closes #23555
closes #23556

split off from #23167
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1254,6 +1254,7 @@ Categorical
 - Bug in :meth:`Categorical.take` with a user-provided ``fill_value`` not encoding the ``fill_value``, which could result in a ``ValueError``, incorrect results, or a segmentation fault (:issue:`23296`).
 - In meth:`Series.unstack`, specifying a ``fill_value`` not present in the categories now raises a ``TypeError`` rather than ignoring the ``fill_value`` (:issue:`23284`)
 - Bug when resampling :meth:`Dataframe.resample()` and aggregating on categorical data, the categorical dtype was getting lost. (:issue:`23227`)
+- Bug in the ``__name__`` attribute of several methods of :class:`Series.str`, which were set incorrectly (:issue:`23551`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.rst]: I'm confused...how does this entry relate to your other changes?

File: pandas/tests/test_strings.py [hunk 2]
```diff
@@ -264,6 +261,7 @@ def test_api_per_method(self, box, dtype,
                          + ['mixed', 'mixed-integer'] * mixed_allowed)
 
         if inferred_dtype in allowed_types:
+            # i.a. GH 23555, GH 23556
```
Comment (gfyoung) [inline pandas/tests/test_strings.py]: i.a. = ?

Comment (h-vetinari) [reply pandas/tests/test_strings.py]: inter alia :)

File: doc/source/whatsnew/v0.24.0.rst (hunk 1 above)
Comment (h-vetinari) [reply doc/source/whatsnew/v0.24.0.rst]: Deleted the wrong line when splitting up #23167...

File: doc/source/whatsnew/v0.24.0.rst [hunk 3]
```diff
@@ -1254,6 +1254,7 @@ Categorical
 - Bug in :meth:`Categorical.take` with a user-provided ``fill_value`` not encoding the ``fill_value``, which could result in a ``ValueError``, incorrect results, or a segmentation fault (:issue:`23296`).
 - In meth:`Series.unstack`, specifying a ``fill_value`` not present in the categories now raises a ``TypeError`` rather than ignoring the ``fill_value`` (:issue:`23284`)
 - Bug when resampling :meth:`Dataframe.resample()` and aggregating on categorical data, the categorical dtype was getting lost. (:issue:`23227`)
+- Bug in many methods of the ``.str``-accessor, which always failed on `CategoricalIndex` (:issue:`23555`, :issue:`23556`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: use double backticks

File: pandas/core/strings.py [hunk 4]
```diff
@@ -1854,15 +1854,16 @@ def __iter__(self):
...
 
         # for category, we do the stuff on the categories, so blow it up
         # to the full series again
         # But for some operations, we have to do the stuff on the full values,
         # so make it possible to skip this step as the method already did this
         # before the transformation...
         if use_codes and self._is_categorical:
-            result = take_1d(result, self._orig.cat.codes,
+            # if self._orig is a CategoricalIndex, there is no .cat-accessor
+            result = take_1d(result, Series(self._orig).cat.codes,
```
Comment (jreback) [inline pandas/core/strings.py]: you can use ``copy=False`` here in the Series construction

File: pandas/tests/test_strings.py [hunk 5]
```diff
@@ -264,6 +261,7 @@ def test_api_per_method(self, box, dtype,
                          + ['mixed', 'mixed-integer'] * mixed_allowed)
 
         if inferred_dtype in allowed_types:
+            # inter alia GH 23555, GH 23556
```
Comment (jreback) [inline pandas/tests/test_strings.py]: just say xref

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, small changes.

File: pandas/tests/test_strings.py (hunk 5 above)
Comment (jreback) [reply pandas/tests/test_strings.py]: this one

File: doc/source/whatsnew/v0.24.0.rst [hunk 6]
```diff
@@ -1256,6 +1256,7 @@ Categorical
 - Bug in :meth:`Categorical.take` with a user-provided ``fill_value`` not encoding the ``fill_value``, which could result in a ``ValueError``, incorrect results, or a segmentation fault (:issue:`23296`).
 - In meth:`Series.unstack`, specifying a ``fill_value`` not present in the categories now raises a ``TypeError`` rather than ignoring the ``fill_value`` (:issue:`23284`)
 - Bug when resampling :meth:`Dataframe.resample()` and aggregating on categorical data, the categorical dtype was getting lost. (:issue:`23227`)
+- Bug in many methods of the ``.str``-accessor, which always failed on ``CategoricalIndex`` (:issue:`23555`, :issue:`23556`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: re-word this to say ``CategoricalIndex.str``

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.rst (hunk 6 above)
Comment (h-vetinari) [reply doc/source/whatsnew/v0.24.0.rst]: @jreback 
Hope this is more like what you were after.

Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: close enough

**Review by jreback — APPROVED** [review]

**Comment by h-vetinari** [comment]
@jreback 
Ping on green

**Comment by jreback** [comment]
thanks @h-vetinari

**Pull request merged**